plugins/la_spec.py

```diff
@@ -0,0 +1,298 @@
+            sptms = [s for s in sptms if s is not None]
+            if not sptms:
+                logging.info("%s plugin cannot load as there are no spectrometers",
+                             self.name)
+                return
+        else:
+            logging.info("%s plugin cannot load as the microscope is not a SPARC",
+                         self.name)
+            return
+
+        for sptm in sptms:
+            if len(sptms) <= 1:
+                actname = "LA Spectrum"
+            else:
+                actname = "LA Spectrum with %s" % (sptm.name,)
+            act = functools.partial(self.addSpectrum, name=actname, detector=sptm)
+            stctrl.add_action(actname, act)
+
+
+        # We "patch" the gui.conf.data for our special stream
+        data.STREAM_SETTINGS_CONFIG[LASpectrumSettingsStream] = (
+            OrderedDict((
+                ("polarization", {
+                    # "control_type": odemis.gui.CONTROL_COMBO,
+                }),
+                ("wavelength", {
+                    "tooltip": "Center wavelength of the spectrograph",
+                    "control_type": odemis.gui.CONTROL_FLT,
+                    "range": (0.0, 1900e-9),
+                }),
+                ("grating", {}),
+                ("slit-in", {
+                    "label": "Input slit",
+                    "tooltip": u"Opening size of the spectrograph input slit.\nA wide opening means more light and a worse resolution.",
+                }),
+            ))
+        )
+
+    def addSpectrum(self, name=None, detector=None):
+        logging.debug("Adding spectrum stream for %s", detector.name)
+
+        main_data = self.main_app.main_data
+        stctrl = self._tab.streambar_controller
+
+        spg = stctrl._getAffectingSpectrograph(detector)
+        spec_stream = LASpectrumSettingsStream(
+            name,
+            detector,
+            detector.data,
+            main_data.ebeam,
+            l2=self._lens2,
+            analyzer=main_data.pol_analyzer,
+            sstage=main_data.scan_stage,
+            opm=main_data.opm,
+            detvas=get_local_vas(detector, main_data.hw_settings_config),
+        )
+
+        # Create the equivalent MDStream
+        sem_stream = self._tab.tab_data_model.semStream
+        sem_spec_stream = LASEMSpectrumMDStream("SEM " + name,
+                                                        [sem_stream, spec_stream])
+
+        # TODO: all the axes, including the filter band should be local. The
+        # band should be set to the pass-through by default
+        axes = {"wavelength": spg,
+                "grating": spg,
+                "slit-in": spg,
+               }
+
+        # Also add light filter for the spectrum stream if it affects the detector
+        for fw in (main_data.cl_filter, main_data.light_filter):
+            if fw is None:
+                continue
+            if detector.name in fw.affects.value:
```

Comment:
I think this doesn't work if detector is None

Reply:
It turns out that the whole code doesn't work with None (and it would never be called with None). I've changed the function signature to reflect that.